### ISSUE TITLE
New membership fields

### ISSUE THREAD

**Issue by ftgo** [tendenci/tendenci#478]
I am trying to customize Tendenci to my needs. One of the changes is to include new fields for Membership registration. It would be nice if we could add new fields but there is no form for this just for custom fields, but these cannot have the type defined. :(

The changes I made on code was (for example): 
`file: tendenci/apps/memberships/fixtures/npo_default_memberships.json`
`
    {
        "pk": 124,
        "model": "memberships.membershipappfield",
        "fields": {
            "default_value": "",
            "field_type": "",
            "admin_only": false,
            "description": "",
            "css_class": "",
            "membership_app": 1,
            "required": false,
            "label": "Allergies",
            "position": 124,
            "content_type": null,
            "help_text": "",
            "choices": "",
            "field_name": "allergies",
            "display": true
        }
    },
`

`file: tendenci/apps/memberships/migrations/0001_initial.py`
`('allergies', models.CharField(verbose_name=b'Allergies', max_length=155, verbose_name='Name'))`

`file: tendenci/apps/memberships/templates/memberships/app_fields.json`
`
    {
    "label": "Allergies",
    "field_name": "allergies",
    "display": "true",
    "required": "false",
    "admin_only": "false"
    },
`

Now I am able to see the fields on Membership Edit screen but it is not shown on the Preview and final Add form. What am I missing?

Tendenci 7.1.381; Django 1.8.8; PostgreSQL 9.3; Ubuntu 14.04 (vagrant ubuntu/trusty32 - machine has no vtx feature);

**Comment by jennyq** [comment]
It doesn't work that way because the field "allergies" doesn't exist in the membershipdefault table. What I'd like to suggest is to use the **User Defined** fields when you make your membership application form at admin backend. There are plenty of them. And there is a place you can edit your field type to suit your needs, check out `http://yoursitedomain/admin/memberships/membershipappfield/`

**Comment by ftgo** [comment]
actually I loaded the field on database during load_base_defaults/load_npo_defaults. it is there.

**Comment by jennyq** [comment]
Although the field was loaded to the membershipappfield table, it doesn't exist as a column in the membershipdefault table. The field_name of each entry in membershipappfield table needs to be one of the column name of membershipdefault, profiles, or auth_user tables...

**Comment by ftgo** [comment]
@jennyq the **User Defined** fields are not shown on `http://yoursitedomain/admin/memberships/membershipappfield/`. I had to run: `update memberships_membershipappfield set display = TRUE where field_name LIKE 'ud%'`

**Comment by jennyq** [comment]
That page only shows the fields selected for the forms. You'll need to first select your UD fields to show on your membership form builder.

**Comment by ftgo** [comment]
got it, now I can create the form I need. However, the membership fields are not shown on custom event forms. Is that a way to show it?

**Comment by jennyq** [comment]
Of course not. Why would you want the membership fields to show on custom event forms?

**Comment by ftgo** [comment]
custom event form has a column disabled for that.

what I want is to create events for memberships of my club. it is natural
to give such a kind of feature, don't you think?

**Comment by jennyq** [comment]
@ftgo, yes you can map some fields to the users fields (in the users/profiles tables) by selecting the checkboxes right above the _Fields_ section. When you select a field, it will automatically map it to a user field. That's why the _Map to User Field_ column is disabled to avoid user error. You can create a new custom registration form if the default one doesn't work.

**Comment by ftgo** [comment]
@jennyq, I am not getting it. Let me see:
- _Users_ are different than _Memberships_ (but they are in the same listings);
- _Users_ form fields are not customizable (translatable?), thus neither _Event_ fields.

If I am right, what can I do to merge _Memberships_ fields with _Users_ on Events?

**Comment by ftgo** [comment]
@jennyq putting this in another way: is it possible to modify _User_ fields?

**Comment by jennyq** [comment]
@ftgo, At the database level, the relationship between users and memberships is a one-to-many relationships. There is a foreign key field user in memberships that points to the auth_user table. You can certainly change the **labels** of the _user_ fields in membership application forms or custom event registration forms, but you cannot change the field names (and you don't need to). Does that make sens

**Comment by eschipul** [comment]
Gonna jump in here. Some history:
users_auth - sacred to django
profiles - our extension of it prior to the ability to create your own user backend, but OK with me.
user-defined-fields - elements from the devil himself brought down to ruin lives and scratch the pain on passing Bentleys for no reason.

The x+4 rule - I have found that if you build a model for a purpose it should accommodate 99% of the use cases and that you should leave open four (4) user defined fields in the FLAT DB table. Then review what clients did with those UD fields every year and debate adding them to the base model (profiles in this case) or creating a new object.

What drives me crazy is user defined data in custom fields in an EAV model that is a vertical enhancement of a fundamentally flat object. That gets pretty meta but let's just say it looks like this, skipping users and just looking at profiles:

Flat table - profiles     firstname   surname    city   
Vertical table
      display label    dblabel     value    fieldtype   etc.....

that soon looks like
----------
John John | Smith | Oakland

with a complex query to get the rest

shoes | shoe-common-name | shoesize | float | etc
Nike   | Air Jordans                |     12.5     |
and then a bunch of other things here so our object now looks like this crap

blah : blah : blah : blah : blah
foo : bar
foo2 : bar2
foo3 : bar3
foo4 : bar4
--------------------
All of this begets a super slow, as if django wasn't slow enough (don't say the word cache please) we get these awkward data structures.

Solution?  Profile the user and have a new object which is an extended profile. This then becomes something like

user_auth ----- profiles ------ persona_type
bob -------------- basics ------- hockeyplayer_data
mary ------------- basics ------- news-anchor-data

Basically even if we have to keep redundant data in sync, flat files from the mainframe days are far more efficient than a strict ORM relational model which deletes related data (like event registrations) when you delete a user. This is why we use statusdetail and status to do a soft-delete on users so we don't delete 50 pages of press releases when you change PR firms.

In closing - YOU ARE RIGHT. It's just a matter of time, money, will, and dialog to get the right db structure.
